### ISSUE TITLE
WebSharper Suave-hosted site -- missing support for static files?

### ISSUE THREAD

**Issue by nightroman** [dotnet-websharper/suave#12]
I am playing with WebSharper 4 but the issue is not necessarily new.

Perhaps I miss something. In a WebSharper 4 Suave-hosted site project created
from the template, I cannot find a way to link static files (css) in my main
html page. The Chrome debugger shows errors on loading such files.

Compare: if I create the similar Owin-hosted site project from the template
then I can link static files (css) right away and all works as expected.

I noticed that Owin-host code has some explicit mention of static files:

```fsharp
use server = WebApp.Start(url, fun appB ->
    appB.UseStaticFiles(
            StaticFileOptions(
                FileSystem = PhysicalFileSystem(rootDirectory)))
        .UseSitelet(rootDirectory, Site.Main)
    |> ignore)
```

I cannot find any Suave-host options related to static files. If it is my bad,
please forgive me and point me in the right direction. I am a novice.

I found a similar question for Suave itself. The answer suggests using
`Files.browseHome`, see [stackoverflow](https://stackoverflow.com/a/35058060/323582).

Then I made a naive hack in [WebSharper.Suave/Main.fs](https://github.com/dotnet-websharper/suave/blob/master/WebSharper.Suave/Main.fs):

```fsharp
match defaultArg ServeResources true with
| true ->
    choose [
        subPath "Scripts"
        subPath "Content"
        Files.browseHome // <- this is the important part
        siteletWebPart
    ]
```

and my static files started to work. I am not suggesting this as a fix. But if
this makes sense I would expect a configuration option which tells
to use this handler among the others.

---

How to repro

- Visual Studio 15.6.4
- Visual Studio WebSharper extension 4.2.8.227
- New project from Visual F# \ WebSharper \ WebSharper 4 Suave-hosted site
- Edit `Main.html` and move its inline styles to the new file, say, `css/app.css`
- Add the link `<link rel="stylesheet" href="/css/app.css"/>` to `Main.html`
- Build
- Run
- Open the browser (Chrome) and navigate to "http://127.0.0.1:8080"

As a result, the main page is shown with some styles missing and the Chrome console log shows
errors on loading `app.css`.

**Comment by Tarmil** [comment]
As an alternative to hacking into WebSharper.Suave, you can use the `Continuation` parameter:

```fsharp
WebSharperAdapter.ToWebPart(Main, RootDirectory = "../..", Continuation = Suave.Files.browseHome)
```

I think this is better than adding an ad-hoc extra configuration option, which will always end up being too narrow for various people's use cases. So I think we shouldn't add anything to WebSharper.Suave itself. However, it would be nice to have it enabled in the default project template. We should replace this:

```fsharp
    do
        startWebServer
            defaultConfig
            (WebSharperAdapter.ToWebPart(Main, RootDirectory = @"../.."))
```

with something like this:

```fsharp
    /// Add here any other Suave WebParts that need to be served, besides WebSharper.
    let suaveSite =
        Suave.Files.browseHome

    do
        let rootDir = "../.."
        startWebServer
            { defaultConfig with homeFolder = Some (System.IO.Path.GetFullPath rootDir) }
            (WebSharperAdapter.ToWebPart(Main, RootDirectory = rootDir, Continuation = suaveSite))
```

**Comment by nightroman** [comment]
@Tarmil Excellent! I knew I might miss some existing and "legit" way.

> However, it would be nice to have it enabled in the default project template.

I think so, too. That would be really useful. Otherwise, at least for a casual user like me, it's a time-consuming problem (or even worse, some users may just give up...).

**Comment by nightroman** [comment]
@Tarmil Did you try this? It does not work or I am missing something again. Note that in my hack it is important to put `Files.browseHome` **before** `siteletWebPart`. So it is not exactly a "continuation" perhaps.

**Comment by nightroman** [comment]
Scratch this, sorry. I missed some bits. It seems to be working!

**Comment by Tarmil** [comment]
I did get this working, yes. Did you make sure that Suave's `homeFolder` is configured correctly (like in my suggested project template change)?

**Comment by nightroman** [comment]
BTW, with this change I cannot repro this problem:
https://github.com/dotnet-websharper/core/issues/933

Can you?

**Comment by Tarmil** [comment]
Indeed, I cannot repro either. I really don't understand why `Continuation` changes anything about that issue though. The only thing changed by `Continuation` is that we call `OwinApp.ofAppWithContinuation` instead of `OwinApp.ofApp`, but [the latter is defined in terms of the former](https://github.com/SuaveIO/suave/blob/51a8f3120a0f7fb5e6e772e8ee703d2e6242ceea/src/Suave/Owin.fs#L766-L772)...

**Comment by nightroman** [comment]
> The only thing changed by Continuation is that we call OwinApp.ofAppWithContinuation instead of OwinApp.ofApp

Yes, but our new continuation is `Suave.Files.browseHome` instead of old default `RequestErrors.NOT_FOUND "File not found"`.

**Comment by Tarmil** [comment]
Sure, but then why does their continuation get erroneously called on `/about` instead of the sitelet, whereas ours correctly isn't?

**Comment by Tarmil** [comment]
I added [this code](https://github.com/dotnet-websharper/suave/issues/12#issuecomment-378239254) to the project template in v4.2.9.256, so this can now be considered fixed. Discussions about whether this fixes dotnet-websharper/core#933 should continue on that ticket.
